diff -uNr WiringPi2-Python_orig/WiringPi/wiringPi/wiringPi.c WiringPi2-Python/WiringPi/wiringPi/wiringPi.c
--- WiringPi2-Python_orig/WiringPi/wiringPi/wiringPi.c	2014-08-04 22:20:53.862804951 -0500
+++ WiringPi2-Python/WiringPi/wiringPi/wiringPi.c	2014-08-05 10:33:04.266995309 -0500
@@ -223,6 +223,7 @@
 // ISR Data
 
 static void (*isrFunctions [64])(void) ;
+static void (*isrFunctionsWithArg [64])(void *) ;
 
 
 // Doing it the Arduino way with lookup tables...
@@ -1400,6 +1401,130 @@
     while (pinPass != -1)
       delay (1) ;
   pthread_mutex_unlock (&pinMutex) ;
+
+  return 0 ;
+}
+
+
+/*
+ * interruptHandlerWithArg:
+ *	This is a thread and gets started to wait for the interrupt we're
+ *	hoping to catch. It will call the user-function when the interrupt
+ *	fires.
+ *********************************************************************************
+ */
+
+static void *interruptHandlerWithArg (void *arg)
+{
+  int myPin ;
+
+  (void)piHiPri (55) ;	// Only effective if we run as root
+
+  myPin   = pinPass ;
+  pinPass = -1 ;
+
+  for (;;)
+    if (waitForInterrupt (myPin, -1) > 0)
+      isrFunctionsWithArg [myPin] (arg) ;
+
+  return NULL ;
+}
+
+
+/*
+ * wiringPiISR:
+ *	Pi Specific.
+ *	Take the details and create an interrupt handler that will do a call-
+ *	back to the user supplied function.
+ *********************************************************************************
+ */
+
+int wiringPiISRWithArg (int pin, int mode, void (*function)(void *), void *arg)
+{
+  pthread_t threadId ;
+  const char *modeS ;
+  char fName   [64] ;
+  char  pinS [8] ;
+  pid_t pid ;
+  int   count, i ;
+  char  c ;
+  int   bcmGpioPin ;
+
+  if ((pin < 0) || (pin > 63))
+    return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: pin must be 0-63 (%d)\n", pin) ;
+
+  /**/ if (wiringPiMode == WPI_MODE_UNINITIALISED)
+    return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: wiringPi has not been initialised. Unable to continue.\n") ;
+  else if (wiringPiMode == WPI_MODE_PINS)
+    bcmGpioPin = pinToGpio [pin] ;
+  else if (wiringPiMode == WPI_MODE_PHYS)
+    bcmGpioPin = physToGpio [pin] ;
+  else
+    bcmGpioPin = pin ;
+
+// Now export the pin and set the right edge
+//	We're going to use the gpio program to do this, so it assumes
+//	a full installation of wiringPi. It's a bit 'clunky', but it
+//	is a way that will work when we're running in "Sys" mode, as
+//	a non-root user. (without sudo)
+
+  if (mode != INT_EDGE_SETUP)
+  {
+    /**/ if (mode == INT_EDGE_FALLING)
+      modeS = "falling" ;
+    else if (mode == INT_EDGE_RISING)
+      modeS = "rising" ;
+    else
+      modeS = "both" ;
+
+    sprintf (pinS, "%d", bcmGpioPin) ;
+
+    if ((pid = fork ()) < 0)	// Fail
+      return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: fork failed: %s\n", strerror (errno)) ;
+
+    if (pid == 0)	// Child, exec
+    {
+      /**/ if (access ("/usr/local/bin/gpio", X_OK) == 0)
+      {
+	execl ("/usr/local/bin/gpio", "gpio", "edge", pinS, modeS, (char *)NULL) ;
+	return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: execl failed: %s\n", strerror (errno)) ;
+      }
+      else if (access ("/usr/bin/gpio", X_OK) == 0)
+      {
+	execl ("/usr/bin/gpio", "gpio", "edge", pinS, modeS, (char *)NULL) ;
+	return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: execl failed: %s\n", strerror (errno)) ;
+      }
+      else
+	return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: Can't find gpio program\n") ;
+    }
+    else		// Parent, wait
+      wait (NULL) ;
+  }
+
+// Now pre-open the /sys/class node - but it may already be open if
+//	we are in Sys mode...
+
+  if (sysFds [bcmGpioPin] == -1)
+  {
+    sprintf (fName, "/sys/class/gpio/gpio%d/value", bcmGpioPin) ;
+    if ((sysFds [bcmGpioPin] = open (fName, O_RDWR)) < 0)
+      return wiringPiFailure (WPI_FATAL, "wiringPiISRWithArg: unable to open %s: %s\n", fName, strerror (errno)) ;
+  }
+
+// Clear any initial pending interrupt
+
+  ioctl (sysFds [bcmGpioPin], FIONREAD, &count) ;
+  for (i = 0 ; i < count ; ++i)
+    read (sysFds [bcmGpioPin], &c, 1) ;
+
+  isrFunctionsWithArg [pin] = function ;
+
+  pthread_mutex_lock (&pinMutex) ;
+    pinPass = pin ;
+    pthread_create (&threadId, NULL, interruptHandlerWithArg, arg) ;
+    while (pinPass != -1)
+      delay (1) ;
+  pthread_mutex_unlock (&pinMutex) ;
 
   return 0 ;
 }
diff -uNr WiringPi2-Python_orig/WiringPi/wiringPi/wiringPi.h WiringPi2-Python/WiringPi/wiringPi/wiringPi.h
--- WiringPi2-Python_orig/WiringPi/wiringPi/wiringPi.h	2014-08-04 22:20:53.862804951 -0500
+++ WiringPi2-Python/WiringPi/wiringPi/wiringPi.h	2014-08-05 10:33:04.266995309 -0500
@@ -165,6 +165,7 @@
 
 extern int  waitForInterrupt    (int pin, int mS) ;
 extern int  wiringPiISR         (int pin, int mode, void (*function)(void)) ;
+extern int  wiringPiISRWithArg  (int pin, int mode, void (*function)(void *), void *arg) ;
 
 // Threads
 
diff -uNr WiringPi2-Python_orig/wiringpi_wrap.c WiringPi2-Python/wiringpi_wrap.c
--- WiringPi2-Python_orig/wiringpi_wrap.c	2014-08-04 22:20:53.862804951 -0500
+++ WiringPi2-Python/wiringpi_wrap.c	2014-08-08 12:38:04.611697151 -0500
@@ -3841,12 +3841,29 @@
   return pyobj;
 }
 
+PyObject *event_callback[64];
+PyObject *event_callback_data[64];
+
+void _wiringPiISR_callback(void *arg) {
+	PyObject *result;
+	int pinNumber = *((int *) arg);
+
+	if (event_callback) {
+		PyGILState_STATE _swig_thread_block = PyGILState_Ensure();
+		result = PyObject_CallFunction(event_callback[pinNumber], "O", event_callback_data[pinNumber]);
+		if (result == NULL && PyErr_Occurred()) {
+			PyErr_Print();
+			PyErr_Clear();
+		}
+		Py_XDECREF(result);
+		PyGILState_Release(_swig_thread_block);
+	}
+}
 
 SWIGINTERN PyObject *_wrap_wiringPiISR(PyObject *SWIGUNUSEDPARM(self), PyObject *args) {
   PyObject *resultobj = 0;
   int arg1 ;
   int arg2 ;
-  void (*arg3)(void) = (void (*)(void)) 0 ;
   int val1 ;
   int ecode1 = 0 ;
   int val2 ;
@@ -3854,9 +3871,12 @@
   PyObject * obj0 = 0 ;
   PyObject * obj1 = 0 ;
   PyObject * obj2 = 0 ;
+  PyObject * cb_func = NULL;
+  PyObject * cb_arg = NULL;
   int result;
+  int *pinNumber;
   
-  if (!PyArg_ParseTuple(args,(char *)"OOO:wiringPiISR",&obj0,&obj1,&obj2)) SWIG_fail;
+  if (!PyArg_ParseTuple(args,(char *)"OOO|O:wiringPiISR",&obj0,&obj1,&obj2,&cb_arg)) SWIG_fail;
   ecode1 = SWIG_AsVal_int(obj0, &val1);
   if (!SWIG_IsOK(ecode1)) {
     SWIG_exception_fail(SWIG_ArgError(ecode1), "in method '" "wiringPiISR" "', argument " "1"" of type '" "int""'");
@@ -3868,12 +3888,26 @@
   } 
   arg2 = (int)(val2);
   {
-    int res = SWIG_ConvertFunctionPtr(obj2, (void**)(&arg3), SWIGTYPE_p_f_void__void);
-    if (!SWIG_IsOK(res)) {
-      SWIG_exception_fail(SWIG_ArgError(res), "in method '" "wiringPiISR" "', argument " "3"" of type '" "void (*)(void)""'"); 
-    }
+	if (!PyArg_ParseTuple(args, "iiO|O", &val1, &val2, &cb_func, &cb_arg)) {
+		SWIG_exception_fail(SWIG_ArgError(SWIG_TypeError), "in method '" "wiringPiISR" "', argument " "3"" of type '" "function""'");
+	}
+	if (!PyCallable_Check(cb_func)) {
+		SWIG_exception_fail(SWIG_ArgError(SWIG_TypeError), "in method '" "wiringPiISR" "', argument " "3"" of type '" "function""'");
+	}
+	if (event_callback[arg1]) {
+		Py_XDECREF(event_callback[arg1]);
+	}
+	if (event_callback_data[arg1]) {
+		Py_XDECREF(event_callback_data[arg1]);
+	}
+    event_callback[arg1] = cb_func;
+    event_callback_data[arg1] = cb_arg;
+    Py_XINCREF(cb_func);
+    Py_XINCREF(cb_arg);
   }
-  result = (int)wiringPiISR(arg1,arg2,arg3);
+  pinNumber = (int *)malloc(sizeof(*pinNumber));
+  *pinNumber = arg1;
+  result = (int)wiringPiISRWithArg(arg1,arg2,&_wiringPiISR_callback,pinNumber);
   resultobj = SWIG_From_int((int)(result));
   return resultobj;
 fail:
